### PR TITLE
Enable pre-rendering of Container addons

### PR DESCRIPTION
**Pull request by tariq1890**
Operation Monkey Brain! :monkey: :brain:

This PR primarily aims are removing the many `sed` statements that we have in our master custom data scripts. Instead of having placeholders that need to be `sed`'ed with ARM parameters, we are looking to use good-old `go text/template` to pre-render the addon files.

1. This PR is intended to be a prototype and we are testing this on the metrics server addon only.
2. We have removed the ARM parameters and `sed` statements that correspond to the metricsServer addon
3. The metrics server manifest YAML now uses go template syntax instead of the `<>` placeholder.

### CODE REVIEW

**Comment by jackfrancis** [comment]
This looks great. Can you write up a brief design doc (one paragraph) summarizing what's going on here.

**Comment by tariq1890** [comment]
Sure!

**Comment by tariq1890** [comment]
Control flow of the add-on pre-render logic.

1. Call template generate
2. Invoke `GetMasterCustomData` template function
3. Invoke newly added `GetK8sAddonString` function inside of `GetMasterCustomData`
4. `GetK8sAddonString` retrieves the go-templated addon manifest files and resolves the templates
5. `GetK8sAddonString` then gzips  and B64's the template-resolved string and creates `cloud-init` write file directives from them.
6. The placeholder `MASTER_ADDONS_NEW_PLACEHOLDER` in `kubernetesmastercustomdata.yml` is then replaced with the generated `cloud-init` write file directives.

File: parts/k8s/containeraddons/kubernetesmasteraddons-aci-connector-deployment.yaml [hunk 1]
```diff
@@ -83,7 +83,7 @@ spec:
         beta.kubernetes.io/os: linux
       containers:
       - name: aci-connector
-        image: <img>
+        image: {{ContaineImage "aci-connector"}}
```
Comment (jackfrancis) [inline parts/k8s/containeraddons/kubernetesmasteraddons-aci-connector-deployment.yaml]: `s/ContaineImage/ContainerImage/`

File: parts/k8s/containeraddons/kubernetesmasteraddons-blobfuse-flexvolume-installer.yaml [hunk 2]
```diff
@@ -29,11 +29,11 @@ spec:
         imagePullPolicy: Always
         resources:
           requests:
-            cpu: <cpuReq>
```
Comment (jackfrancis) [inline parts/k8s/containeraddons/kubernetesmasteraddons-blobfuse-flexvolume-installer.yaml]: While we're doing this change, can we remove the static `andyzhangx/blobfuse-flexvol-installer` config (above) and wire it up for `ContainerImage` template func usage?

File: parts/k8s/containeraddons/kubernetesmasteraddons-smb-flexvolume-installer.yaml [hunk 3]
```diff
@@ -29,11 +29,11 @@ spec:
         imagePullPolicy: Always
         resources:
           requests:
-            cpu: <cpuReq>
```
Comment (jackfrancis) [inline parts/k8s/containeraddons/kubernetesmasteraddons-smb-flexvolume-installer.yaml]: ditto the above command about removing the static image reference

File: parts/k8s/containeraddons/kubernetesmasteraddons-blobfuse-flexvolume-installer.yaml (hunk 2 above)
Comment (tariq1890) [reply parts/k8s/containeraddons/kubernetesmasteraddons-blobfuse-flexvolume-installer.yaml]: Sure!

File: parts/k8s/containeraddons/kubernetesmasteraddons-nvidia-device-plugin-daemonset.yaml [hunk 4]
```diff
@@ -38,15 +38,15 @@ spec:
...
         resources:
           requests:
-            cpu: <cpuReq>
-            memory: <memReq>
+            cpu: image: {{ContainerCPUReqs "nvidia-device-plugin"}}
+            memory: image: {{ContainerMemReqs "nvidia-device-plugin"}}
           limits:
-            cpu: <cpuLim>
-            memory: <memLim>
+            cpu: image: {{ContainerCPULimits "nvidia-device-plugin"}}
```
Comment (CecileRobertMichon) [inline parts/k8s/containeraddons/kubernetesmasteraddons-nvidia-device-plugin-daemonset.yaml]: why the extra "image:" here?

File: pkg/api/types.go [hunk 5]
```diff
@@ -1235,7 +1235,7 @@ func (k *KubernetesConfig) IsAADPodIdentityEnabled() bool {
 
 // IsACIConnectorEnabled checks if the ACI Connector addon is enabled
 func (k *KubernetesConfig) IsACIConnectorEnabled() bool {
-	return k.isAddonEnabled(DefaultACIConnectorAddonName, DefaultAADPodIdentityAddonEnabled)
+	return k.isAddonEnabled(DefaultACIConnectorAddonName, DefaultACIConnectorAddonEnabled)
```
Comment (CecileRobertMichon) [inline pkg/api/types.go]: nice

**Comment by tariq1890** [comment]
Ready for another round of review @CecileRobertMichon @jackfrancis

File: pkg/acsengine/artifacts.go [hunk 6]
```diff
@@ -268,6 +274,19 @@ func kubernetesArtifactSettingsInitAgent(profile *api.Properties) []kubernetesFe
 	}
 }
 
+func getAddonString(input, destinationPath, destinationFile string) string {
```
Comment (CecileRobertMichon) [inline pkg/acsengine/artifacts.go]: can this be unit tested?

File: parts/k8s/kubernetesmastercustomdata.yml [hunk 7]
```diff
@@ -302,54 +301,18 @@ MASTER_ARTIFACTS_CONFIG_PLACEHOLDER
...
-{{end}}
-
 {{if .OrchestratorProfile.KubernetesConfig.IsClusterAutoscalerEnabled}}
-    sed -i "s|<cloud>|{{WrapAsParameter "kubernetesClusterAutoscalerAzureCloud"}}|g; s|<img>|{{WrapAsParameter "kubernetesClusterAutoscalerSpec"}}|g; s|<cpuLim>|{{WrapAsParameter "kubernetesClusterAutoscalerCPULimit"}}|g; s|<memLim>|{{WrapAsParameter "kubernetesClusterAutoscalerMemoryLimit"}}|g; s|<cpuReq>|{{WrapAsParameter "kubernetesClusterAutoscalerCPURequests"}}|g; s|<memReq>|{{WrapAsParameter "kubernetesClusterAutoscalerMemoryRequests"}}|g; s|<minNodes>|{{WrapAsParameter "kubernetesClusterAutoscalerMinNodes"}}|g; s|<maxNodes>|{{WrapAsParameter "kubernetesClusterAutoscalerMaxNodes"}}|g; s|<mie>|{{WrapAsParameter "kubernetesClusterAutoscalerUseManagedIdentity"}}|g" /etc/kubernetes/addons/cluster-autoscaler-deployment.yaml
+    sed -i "s|<cloud>|{{WrapAsParameter "kubernetesClusterAutoscalerAzureCloud"}}|g; s|<mie>|{{WrapAsParameter "kubernetesClusterAutoscalerUseManagedIdentity"}}|g" /etc/kubernetes/addons/cluster-autoscaler-deployment.yaml
```
Comment (CecileRobertMichon) [inline parts/k8s/kubernetesmastercustomdata.yml]: would it be possible to change `mie` back to something more verbose now that we have more custom data space?

File: parts/k8s/kubernetesmastercustomdata.yml [hunk 8]
```diff
@@ -386,12 +349,9 @@ MASTER_ARTIFACTS_CONFIG_PLACEHOLDER
 {{end}}
 {{if HasCustomSearchDomain}}
     sed -i "s|<searchDomainName>|{{WrapAsParameter "searchDomainName"}}|g; s|<searchDomainRealmUser>|{{WrapAsParameter "searchDomainRealmUser"}}|g; s|<searchDomainRealmPassword>|{{WrapAsParameter "searchDomainRealmPassword"}}|g" /opt/azure/containers/setup-custom-search-domains.sh
-{{end}}
-{{if .OrchestratorProfile.KubernetesConfig.IsContainerMonitoringEnabled}}
-    sed -i "s|<ver>|{{WrapAsParameter "omsAgentVersion"}}|g; s|<dockerVer>|{{WrapAsParameter "omsAgentDockerProviderVersion"}}|g; s|<img>|{{WrapAsParameter "omsAgentImage"}}|g; s|<wsid>|{{WrapAsParameter "omsAgentWorkspaceGuid"}}|g; s|<wsKey>|{{WrapAsParameter "omsAgentWorkspaceKey"}}|g; s|<cpuReq>|{{WrapAsParameter "kubernetesOMSAgentCPURequests"}}|g; s|<memReq>|{{WrapAsParameter "kubernetesOMSAgentMemoryRequests"}}|g; s|<cpuLim>|{{WrapAsParameter "kubernetesOMSAgentCPULimit"}}|g; s|<memLim>|{{WrapAsParameter "kubernetesOMSAgentMemoryLimit"}}|g" /etc/kubernetes/addons/omsagent-daemonset.yaml
 {{end}}
     a=/etc/kubernetes/addons/ip-masq-agent.yaml
```
Comment (CecileRobertMichon) [inline parts/k8s/kubernetesmastercustomdata.yml]: We could probably get rid of this var definition now

File: pkg/acsengine/artifacts.go (hunk 6 above)
Comment (tariq1890) [reply pkg/acsengine/artifacts.go]: screen shot 2018-11-13 at 12 16 36 pm

I've run tests and seen that the method is already covered

File: parts/k8s/kubernetesmastercustomdata.yml (hunk 8 above)
Comment (tariq1890) [reply parts/k8s/kubernetesmastercustomdata.yml]: Not sure I understand.

File: parts/k8s/kubernetesmastercustomdata.yml (hunk 7 above)
Comment (tariq1890) [reply parts/k8s/kubernetesmastercustomdata.yml]: Sure!

File: parts/k8s/kubernetesmastercustomdata.yml (hunk 8 above)
Comment (CecileRobertMichon) [reply parts/k8s/kubernetesmastercustomdata.yml]: it was added as a quick way to save custom data space in @jackfrancis'  PR but if we're not as tight on space after this PR it's not great in terms of code readability

File: pkg/acsengine/artifacts.go (hunk 6 above)
Comment (CecileRobertMichon) [reply pkg/acsengine/artifacts.go]: great

File: parts/k8s/containeraddons/kubernetesmasteraddons-aci-connector-deployment.yaml (hunk 1 above)
Comment (tariq1890) [reply parts/k8s/containeraddons/kubernetesmasteraddons-aci-connector-deployment.yaml]: Done

File: parts/k8s/kubernetesmastercustomdata.yml (hunk 8 above)
Comment (tariq1890) [reply parts/k8s/kubernetesmastercustomdata.yml]: The var `a` is used in multiple lines.

**Pull request merged**